src/Raven.Server/Documents/Handlers/AI/AiAgent/AiAgentHandler.cs

```diff
@@ -0,0 +1,415 @@
+        using (ServerStore.Engine.ContextPool.AllocateOperationContext(out ClusterOperationContext ctx))
+        using (ctx.OpenReadTransaction())
+        using (var record = ServerStore.Cluster.ReadRawDatabaseRecord(ctx, DatabaseName))
+        {
+            agents = record.AiAgents;
+        }
+
+        using (ServerStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+        await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream(), token.Token))
+        {
+            writer.WriteStartObject();
+
+            var count = agents.Count;
+            var current = 0;
+            foreach (var agent in agents)
+            {
+                current++;
+                writer.WritePropertyName(agent.Key);
+                writer.WriteObject(context.ReadObject(agent.Value.ToJson(), "ai-agent"));
+                if (current < count)
+                {
+                    writer.WriteComma();
+                }
+            }
+            writer.WriteEndObject();
+        }
+    }
+
+    [RavenAction("/databases/*/ai/ai-agent/test", "POST", AuthorizationStatus.ValidUser, EndpointType.Read, DisableOnCpuCreditsExhaustion = true)]
+    public async Task AiAgentTest()
+    {
+        using var _ = ContextPool.AllocateOperationContext(out DocumentsOperationContext context);
+        using var token = CreateHttpRequestBoundOperationToken();
+        string userPrompt = GetStringQueryString("prompt", required: true);
+        var options = await context.ReadForMemoryAsync(RequestBodyStream(), "ai-agent", token.Token);
+        var cfg = JsonDeserializationClient.AiAgentConfiguration(options);
+        options.TryGet("Parameters", out BlittableJsonReaderObject parameters);
+
+        var r = await Talk(context, cfg, userPrompt, parameters, token);
+
+        await using var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream());
+        writer.WriteStartObject();
+        writer.WritePropertyName("Result");
+        writer.WriteObject(r.Response.Result);
+        writer.WriteComma();
+        writer.WritePropertyName("Usage");
+        r.Usage.Write(writer);
+        writer.WriteEndObject();
+    }
+
+    private async Task<(AiUsage Usage, AiResponse Response, BlittableJsonReaderObject Dcoument)> Talk(DocumentsOperationContext context, AiAgentConfiguration cfg,
+        string userPrompt, BlittableJsonReaderObject parameters, OperationCancelToken token)
+    {
+        var conStr = GetAiConnectionString(cfg.ConnectionStringName);
+
+        string schemaOrSampleObject = cfg.OutputSchema ?? throw new InvalidOperationException("Missing output schema in configuration");
+        string schema = ChatCompletionClient.GetSchemaFor(schemaOrSampleObject);
+        using var client = ChatCompletionClient.CreateChatCompletionClient(Database.ServerStore.ContextPool, conStr, schema);
+
+        List<BlittableJsonReaderObject> msgs =
```

Comment:
Note that this impl works for _starting_ a conversation - but this won't work for resuming. You need to get the messages from outside.

Reply:
yes, will be handled under different issue https://issues.hibernatingrhinos.com/issue/RavenDB-24208